### PR TITLE
Add CumSum and Round for Opset 11

### PR DESCRIPTION
**Pull request by jignparm**
**Description**: CumSum() and Round() are new operators for ONNX opset 11

**Motivation and Context**
Add implementations of new operators to OnnxRuntime.

### CODE REVIEW

File: onnxruntime/test/onnx/main.cc [hunk 1]
```diff
@@ -404,7 +404,7 @@ int real_main(int argc, char* argv[], Ort::Env& env) {
       {"cumsum_1d_reverse_exclusive", "not implemented yet"},
       {"cumsum_1d_reverse", "not implemented yet"},
       {"cumsum_1d_exclusive", "not implemented yet"},
-      {"cumsum_1d", "not implemented yet"},      
+      {"cumsum_1d", "not implemented yet"},
```
[the remark this answers is not in the document]
Comment (jignparm) [reply onnxruntime/test/onnx/main.cc]: Removed these tests.  The 'cumsum_1d_reverse_exclusive' test is still blacklisted, due to failure in Linux GPU CI build. The likely cause is due to bad ONNX test data in a previous commit ID (CI tests on multiple versions of ONNX test data). However, the op itself is fine, and the ONNX test data should be correct in onnxtip tag.

File: onnxruntime/test/python/onnx_backend_test_series.py [hunk 2]
```diff
@@ -103,13 +103,6 @@ def create_backend_test(testname=None):
...
                                  '^test_bitshift_left_uint8_cpu.*',
-                                 '^test_round_cpu.*',
-                                 '^test_cumsum_1d_cpu.*',
-                                 '^test_cumsum_1d_exclusive_cpu.*',
-                                 '^test_cumsum_1d_reverse_cpu.*',
-                                 '^test_cumsum_1d_reverse_exclusive_cpu.*',
-                                 '^test_cumsum_2d_axis_0_cpu.*',
-                                 '^test_cumsum_2d_axis_1_cpu.*',
                                  '^test_dynamicquantizelinear*',
                                  '^test_dynamicquantizelinear_expanded*',
```
[the remark this answers is not in the document]
Comment (jignparm) [reply onnxruntime/test/python/onnx_backend_test_series.py]: Removed the round and cumsum tests. Seems like the '^test_dynamicquantizelinear_*' tests require additional opset11 ops (in addition to round) to pass successfully, so they are still left as blacklisted.

File: onnxruntime/core/providers/cpu/cpu_execution_provider.cc [hunk 3]
```diff
@@ -302,6 +302,19 @@ class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 10, Re
 
 // opset 11
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 11, Clip);
+class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 11, float, CumSum);
```
[the remark this answers is not in the document]
Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/cpu_execution_provider.cc]: +1. Let's start with float only. If needed other types can be added later on.

File: onnxruntime/core/providers/cpu/math/cumsum.cc [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                const std::vector<int64_t>& slice_dims, const std::vector<int64_t>& steps, const int64_t slice_size) {
+  SliceIterator<T> input_iterator(input, input_starts, slice_dims, steps);
+  WritableSliceIterator<T> output_iterator(output, output_starts, slice_dims, steps);
+  for (int64_t k = 0; k < slice_size; ++k, ++output_iterator, ++input_iterator) {
+    *output_iterator = *input_iterator;
+  }
+}
+template <typename T>
+void SumSlices(const Tensor& input, Tensor& output,
+               std::vector<int64_t>& input_starts, const std::vector<int64_t>& output_starts, const std::vector<int64_t>& previous_output_starts,
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/math/cumsum.cc]: can input_starts be passed by const-ref?

File: onnxruntime/core/providers/cpu/math/cumsum.h [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+template <class T>
+class CumSum final : public OpKernel {
+ public:
+  explicit CumSum(const OpKernelInfo& op_kernel_info);
+
+  Status Compute(OpKernelContext* p_op_kernel_context) const override;
+
+ private:
+  int64_t _exclusive;
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/math/cumsum.h]: exclusive_ and reverse_ (naming convention - trailing underscore)

File: onnxruntime/core/providers/cpu/math/round.cc [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+namespace onnxruntime {
+
+ONNX_CPU_OPERATOR_TYPED_KERNEL(Round, 11, MLFloat16, KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<MLFloat16>()), Round<MLFloat16>);
+ONNX_CPU_OPERATOR_TYPED_KERNEL(Round, 11, float, KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<float>()), Round<float>);
+ONNX_CPU_OPERATOR_TYPED_KERNEL(Round, 11, double, KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<double>()), Round<double>);
+
+
+template <typename T>
+Status Round<T>::Compute(OpKernelContext* ctx) const {
+  auto& X = *ctx->Input<Tensor>(0);
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/math/round.cc]: please prefix auto with const where applicable. See [here](https://github.com/microsoft/onnxruntime/blob/master/docs/Coding_Conventions_and_Standards.md)

File: onnxruntime/core/providers/cpu/tensor/utils.h [hunk 7]
```diff
@@ -277,4 +277,115 @@ inline void CopyCpuTensor(const Tensor* src, Tensor* tgt) {
...
+                        gsl::span<const int64_t> extents, gsl::span<const int64_t> steps)
+      : tensor_(tensor), input_(tensor_.template MutableData<T>()), extents_(extents), skips_(tensor_shape, extents, steps), indices_(extents.size(), 0) {
+    auto& dims = tensor_shape.GetDims();
+    Init(dims, starts, steps);
+  }
+
+  // Initialize initial skip and inner_extent.
+  void Init(const std::vector<int64_t>& dims, gsl::span<const int64_t> starts,
+            gsl::span<const int64_t> steps) {
+    ORT_ENFORCE(static_cast<ptrdiff_t>(dims.size()) == starts.size() &&
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/utils.h]: Append a message to ORT_ENFORCEs for better debuggability.

File: onnxruntime/core/providers/cpu/cpu_execution_provider.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (jignparm) [reply onnxruntime/core/providers/cpu/cpu_execution_provider.cc]: Done- keeping only float, int32 and int64. Removed other uncommon types.

File: onnxruntime/core/providers/cpu/math/cumsum.cc (hunk 4 above)
Comment (jignparm) [reply onnxruntime/core/providers/cpu/math/cumsum.cc]: > can input_starts be passed by const-ref?

Modified to const by ref.

File: onnxruntime/core/providers/cpu/math/cumsum.h (hunk 5 above)
Comment (jignparm) [reply onnxruntime/core/providers/cpu/math/cumsum.h]: > exclusive_ and reverse_ (naming convention - trailing underscore)

Renamed with trailing underscore.

File: onnxruntime/core/providers/cpu/math/round.cc (hunk 6 above)
Comment (jignparm) [reply onnxruntime/core/providers/cpu/math/round.cc]: > please prefix auto with const where applicable. See [here](https://github.com/microsoft/onnxruntime/blob/master/docs/Coding_Conventions_and_Standards.md)

Updated with const auto& where applicable.

File: onnxruntime/core/providers/cpu/tensor/utils.h (hunk 7 above)
Comment (jignparm) [reply onnxruntime/core/providers/cpu/tensor/utils.h]: > Append a message to ORT_ENFORCEs for better debuggability.

Added friendlier error message for these conditions separately.

File: onnxruntime/core/providers/cpu/math/round.cc [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  for (int64_t i = 0; i < size; ++i, ++output, ++input) {
+    *output = ::rint(*input);
+  }
+  return Status::OK();
+}
+template <>
+Status Round<MLFloat16>::Compute(OpKernelContext* ctx) const {
+  const auto& X = *ctx->Input<Tensor>(0);
+  auto& Y = *ctx->Output(0, X.Shape());
+  auto input = X.template Data<MLFloat16>();
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/math/round.cc]: There are several places (other than this line) where RHS is of type ptr and LHS is just auto. As per the coding guidelines we should "Qualify usages of 'auto' with 'const', '*', '&' and '&&' where applicable to more clearly express the intent".

Comment (jignparm) [reply onnxruntime/core/providers/cpu/math/round.cc]: Updated in the other locations as well, to make it more clear which auto vars are consts and pointers.

File: onnxruntime/core/providers/cpu/tensor/utils.h [hunk 9]
```diff
@@ -277,4 +277,120 @@ inline void CopyCpuTensor(const Tensor* src, Tensor* tgt) {
...
+  }
+
+  // This construct takes a explicit tensor_shape which might be different from the shape defined in input tensor.
+  // The explicit tensor_shape usually has inner most axis flattened. For example, given shape[1,4,4,2], if last axis
+  // does not have padding or slice, then it will be flattened as [1,4,8] for better performance (One inner most copy instead of 4).
+  // Also supports arbitrary positive and negative stepping along individual axes
+  WritableSliceIterator(Tensor& tensor, const TensorShape& tensor_shape, gsl::span<const int64_t> starts,
+                        gsl::span<const int64_t> extents, gsl::span<const int64_t> steps)
+      : tensor_(tensor), input_(tensor_.template MutableData<T>()), extents_(extents), skips_(tensor_shape, extents, steps), indices_(extents.size(), 0) {
+    auto& dims = tensor_shape.GetDims();
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/utils.h]: const auto&

File: onnxruntime/core/providers/cpu/tensor/utils.h [hunk 10]
```diff
@@ -277,4 +277,120 @@ inline void CopyCpuTensor(const Tensor* src, Tensor* tgt) {
   }
 }
 
+// This provides easy sequential iteration over a subset of a tensor given a span of starts, extents & optionally steps
+template <typename T>
+struct WritableSliceIterator {
+  WritableSliceIterator(Tensor& tensor, gsl::span<const int64_t> starts,
+                        gsl::span<const int64_t> extents, gsl::span<const int64_t> steps)
+      : tensor_(tensor), input_(tensor_.template MutableData<T>()), extents_(extents), skips_(tensor_.Shape(), extents, steps), indices_(extents.size(), 0) {
+    auto& dims = tensor_.Shape().GetDims();
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/utils.h]: const auto&

**Pull request merged**